# Exercise 5 <br>
Student: Duncan Ferguson <br>
Student Id: 871641260 <br>
Class: Comp 4431-1 <br>
Assignment: Exercise 5 <br>
Date: 10/16/2021 <br>
Group: Name: Broken Toe <br>
Group Members: Emma Bright, Mike Santoro <br>

In [1]:
import pandas as pd
from csv import reader
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

pd.set_option('display.max_rows', None)
pd.set_option('display.width', None)

theData = [
['milk','onion','egg','bread','chili'],
['milk','egg','bread','chili'],
['milk','garlic','beans','chili'],
['milk','egg','beans','chili'],
['milk','onion','egg','bread'],
['milk','onion','egg'],
['garlic','onion','beans'],
['garlic','onion','beans'],
['garlic','onion','beans'],
['garlic','onion','beans'],
['chili','garlic','onion','beans'],
['chili','garlic','onion','beans'],
['garlic','onion','milk'],
['garlic','onion'],
['garlic','chili'],
['garlic','chili'],
['milk','onion','egg'],
['milk','onion','egg'],
['milk','onion','egg','bread'],
['milk','onion','chili','bread'],
]

# Use Transaction encoder to transform into an 1-hot boolean encoded numpy array
te = TransactionEncoder()
te_ary = te.fit(theData).transform(theData)

print("te_aray returned from Transactionencoder.fit().transform()")
te_ary

te_aray returned from Transactionencoder.fit().transform()


array([[False,  True,  True,  True, False,  True,  True],
       [False,  True,  True,  True, False,  True, False],
       [ True, False,  True, False,  True,  True, False],
       [ True, False,  True,  True, False,  True, False],
       [False,  True, False,  True, False,  True,  True],
       [False, False, False,  True, False,  True,  True],
       [ True, False, False, False,  True, False,  True],
       [ True, False, False, False,  True, False,  True],
       [ True, False, False, False,  True, False,  True],
       [ True, False, False, False,  True, False,  True],
       [ True, False,  True, False,  True, False,  True],
       [ True, False,  True, False,  True, False,  True],
       [False, False, False, False,  True,  True,  True],
       [False, False, False, False,  True, False,  True],
       [False, False,  True, False,  True, False, False],
       [False, False,  True, False,  True, False, False],
       [False, False, False,  True, False,  True,  True],
       [False,

In [2]:
# Convert into a dataframe for convenience and to pas into the apriori
df2 = pd.DataFrame(te_ary, columns=te.columns_)

print("DataFrame Version:")
df2.head(25)

DataFrame Version:


beans  bread  chili    egg  garlic   milk  onion
0   False   True   True   True   False   True   True
1   False   True   True   True   False   True  False
2    True  False   True  False    True   True  False
3    True  False   True   True   False   True  False
4   False   True  False   True   False   True   True
5   False  False  False   True   False   True   True
6    True  False  False  False    True  False   True
7    True  False  False  False    True  False   True
8    True  False  False  False    True  False   True
9    True  False  False  False    True  False   True
10   True  False   True  False    True  False   True
11   True  False   True  False    True  False   True
12  False  False  False  False    True   True   True
13  False  False  False  False    True  False   True
14  False  False   True  False    True  False  False
15  False  False   True  False    True  False  False
16  False  False  False   True   False   True   True
17  False  False  False   True   False   True   True
18  False   True  False   True   False   True   True
19  False   True   True  False   False   True   True

In [3]:
# Call apriori to find frequent itemsets with min_support = 30%
freq_items = apriori(df2, min_support=0.3, use_colnames=True)
freq_items['length'] = freq_items['itemsets'].apply(lambda x: len(x))
freq_items.sort_values(['support', 'length'], inplace=True, ascending=False)
print("Frequent Items:")
freq_items

Frequent Items:


support                itemsets  length
5      0.75                 (onion)       1
3      0.55                (garlic)       1
4      0.55                  (milk)       1
1      0.45                 (chili)       1
8      0.40             (egg, milk)       2
10     0.40         (garlic, onion)       2
11     0.40           (onion, milk)       2
0      0.40                 (beans)       1
2      0.40                   (egg)       1
6      0.35         (beans, garlic)       2
12     0.30  (beans, garlic, onion)       3
13     0.30      (egg, onion, milk)       3
7      0.30          (beans, onion)       2
9      0.30            (egg, onion)       2

In [4]:
# Mining the rules by calling association_rules
print("The Rules")
rules = association_rules(freq_items, metric="confidence", min_threshold=0.7)
rules.columns = ['ante', 'consequ', 'antsup', 'consup', 'sup', 'conf', 'lift', 'leverage', 'convic']
rules[["ante","consequ","antsup","consup","sup","conf","lift"]]

The Rules


ante          consequ  antsup  consup   sup      conf      lift
0             (egg)           (milk)    0.40    0.55  0.40  1.000000  1.818182
1            (milk)            (egg)    0.55    0.40  0.40  0.727273  1.818182
2          (garlic)          (onion)    0.55    0.75  0.40  0.727273  0.969697
3            (milk)          (onion)    0.55    0.75  0.40  0.727273  0.969697
4           (beans)         (garlic)    0.40    0.55  0.35  0.875000  1.590909
5   (beans, garlic)          (onion)    0.35    0.75  0.30  0.857143  1.142857
6    (beans, onion)         (garlic)    0.30    0.55  0.30  1.000000  1.818182
7   (garlic, onion)          (beans)    0.40    0.40  0.30  0.750000  1.875000
8           (beans)  (garlic, onion)    0.40    0.40  0.30  0.750000  1.875000
9      (egg, onion)           (milk)    0.30    0.55  0.30  1.000000  1.818182
10      (egg, milk)          (onion)    0.40    0.75  0.30  0.750000  1.000000
11    (milk, onion)            (egg)    0.40    0.40  0.30  0.750000  1.875000
12            (egg)    (milk, onion)    0.40    0.40  0.30  0.750000  1.875000
13          (beans)          (onion)    0.40    0.75  0.30  0.750000  1.000000
14            (egg)          (onion)    0.40    0.75  0.30  0.750000  1.000000

Questions 1: <br>
Using the values already set in the code  <br>
(i.e. apriori min_support = 0.3, association_rule min_threshold = 0.7) <br>
look at the result an answer: <br>

a). Which 1-itemset has the greatest support? <br>
-onion with a support of 0.75

In [5]:
freq_items.sort_values(['support', 'length'], inplace=True, ascending=False)
a1 = freq_items[((freq_items['length'] == 1) & (freq_items['support'] == freq_items.support.max()))]
a1

support itemsets  length
5     0.75  (onion)       1

b). which 3-itemsets have support >= 0.3
- 0.062745	(sugar, ghee)

In [6]:
b1 = freq_items[((freq_items['length'] == 3) & (freq_items['support'] >= 0.3))]
b1

support                itemsets  length
12      0.3  (beans, garlic, onion)       3
13      0.3      (egg, onion, milk)       3

c). Why do the two rules (milk->egg and egg->milk) have different confidence values? <br>
The reason that (milk->egg and egg->milk) have different confidence values is because they have different
levels of support. There conditional probabilities are different because the frequency (support value)
of the 1-itemset milk and the frequency of the 1-itemset (support value) are different. <br>
-Milk(support) -> 0.55 <br>
-Egg(support) -> 0.40



In [7]:
c1 = freq_items[(freq_items["itemsets"] == frozenset({'milk'})) | (freq_items["itemsets"] == frozenset({'egg'}))]
c1

support itemsets  length
4     0.55   (milk)       1
2     0.40    (egg)       1

d) Which rules are not "interesting"? Why do you say that?

For a rule to be considered interesting it must have a lift that is gerater than 1. The following rules are not
interesting because they either have a lift of 1 (meaning that there is no correlation between the itemsets) or
they have a lift that is less than 1 (meaning the itemsets are negatively correlated)
- Beans -> Onion
- Egg -> Onion
- Garlic -> Onion
- Milk -> Onion
- Milk, Egg -> Onion

In [8]:
not_interesting = rules[["ante","consequ","antsup","consup","sup","conf","lift"]]
d1 = not_interesting[not_interesting["lift"] <= 1]
d1

ante  consequ  antsup  consup  sup      conf      lift
2      (garlic)  (onion)    0.55    0.75  0.4  0.727273  0.969697
3        (milk)  (onion)    0.55    0.75  0.4  0.727273  0.969697
10  (egg, milk)  (onion)    0.40    0.75  0.3  0.750000  1.000000
13      (beans)  (onion)    0.40    0.75  0.3  0.750000  1.000000
14        (egg)  (onion)    0.40    0.75  0.3  0.750000  1.000000

e) Two Rules have a lift value < 1.0 (garlic-> onion and milk->onion). Why are these especially useless rules?

These rules are especially useless rules because they have a lift value of less than 1, indicating that the two
itemsets are negatively correlated. The occurrence of one item will likely lead to the absence of the other item
- Egg -> Onion
- Garlic -> Onion

In [9]:
e1 = not_interesting[not_interesting["lift"] <= 0.99999]
e1

ante  consequ  antsup  consup  sup      conf      lift
2  (garlic)  (onion)    0.55    0.75  0.4  0.727273  0.969697
3    (milk)  (onion)    0.55    0.75  0.4  0.727273  0.969697

### 2) The following file contains a clean version of ingredients in 256 indian dishes <br>
Modify the example code to read in and process this file.

In [10]:
from csv import reader

the_2_data = []
with open('exercise5b_input.csv.txt', 'r') as read_obj:
    csv_reader = reader(read_obj)
    for row in csv_reader:
        the_2_data.append(row)


# Using the Transaction Encoder to Transform into an 1-hot boolean encoded numpy array
te = TransactionEncoder()
te_ary = te.fit(the_2_data).transform(the_2_data)

# Converting into a dataframe for the convenience of passing into apriori
df_2 = pd.DataFrame(te_ary, columns= te.columns_)
df_2.head(25)

jaggery  all purpose flour  almond  almonds   aloo  alum powder  \
0      False              False   False    False  False        False   
1      False              False   False    False  False        False   
2      False              False   False    False  False        False   
3      False              False   False     True  False        False   
4      False              False   False    False  False        False   
5      False              False   False    False  False        False   
6      False              False   False    False  False        False   
7      False              False   False    False  False        False   
8      False              False   False    False  False        False   
9      False              False   False    False  False        False   
10     False              False   False    False  False        False   
11     False              False   False    False  False        False   
12     False              False   False    False  False        False   
13     False              False   False    False  False         True   
14     False              False   False    False  False        False   
15     False              False   False    False  False        False   
16     False              False   False    False  False        False   
17     False              False   False    False  False        False   
18     False              False   False    False  False        False   
19     False              False   False    False  False        False   
20     False              False   False    False  False        False   
21     False              False   False    False  False        False   
22     False              False   False    False  False        False   
23     False              False   False    False  False        False   
24     False              False   False    False  False        False   

    amaranth leaves  amchur powder  apricots  arbi ke patte  ...  \
0             False          False     False          False  ...   
1             False          False     False          False  ...   
2             False          False     False          False  ...   
3             False          False     False          False  ...   
4             False          False     False          False  ...   
5             False          False     False          False  ...   
6             False          False     False          False  ...   
7             False          False     False          False  ...   
8             False          False     False          False  ...   
9             False          False     False          False  ...   
10            False          False     False          False  ...   
11            False          False     False          False  ...   
12            False          False     False          False  ...   
13            False          False     False          False  ...   
14            False          False     False          False  ...   
15            False          False     False          False  ...   
16            False          False     False          False  ...   
17            False          False     False          False  ...   
18            False          False     False          False  ...   
19            False          False     False          False  ...   
20            False          False     False          False  ...   
21            False          False     False          False  ...   
22            False          False     False          False  ...   
23            False          False     False          False  ...   
24            False          False     False          False  ...   

    whole urad dal  whole wheat bread  whole wheat flour  whole wheat rava  \
0            False              False              False             False   
1            False              False              False             False   
2            False              False              False             False   
3            False              False              

Once read in, explore the file for various min_support and min_confidence settings.

Note it is suggested to start exploring with: <br>
min_support = 1% or 2% and min_confidence = 30% <br>
You can then parameter up/down and see what happens.

In [161]:
# Looking at Minsupport of 2%
freq_items_2 = apriori(df_2, min_support=0.02, use_colnames=True)
freq_items_2['length'] = freq_items_2['itemsets'].apply(lambda x: len(x))
freq_items_2.sort_values(["length", "support"], inplace=True, ascending=True)
rules_2 = association_rules(freq_items_2, metric="confidence", min_threshold=0.01)
rules_2_min_conf_3 = rules_2[((rules_2['confidence'] >= 0.3) & (rules_2['support'] >= 0.02))].copy()
rules_2_min_conf_3

antecedents     consequents  antecedent support  consequent support  \
3   (mustard seeds)  (curry leaves)            0.035294            0.098039   
4         (jaggery)          (ghee)            0.074510            0.098039   
6          (tomato)        (ginger)            0.062745            0.113725   
8         (jaggery)    (rice flour)            0.074510            0.082353   
10        (saffron)         (sugar)            0.039216            0.188235   
13       (cardamom)         (sugar)            0.035294            0.188235   
14   (curry leaves)      (urad dal)            0.098039            0.098039   
15       (urad dal)  (curry leaves)            0.098039            0.098039   
16   (garam masala)        (ginger)            0.105882            0.113725   
17         (ginger)  (garam masala)            0.113725            0.105882   
18      (chana dal)      (urad dal)            0.062745            0.098039   
19       (urad dal)     (chana dal)            0.098039            0.062745   
20           (milk)         (sugar)            0.082353            0.188235   
22          (sugar)          (ghee)            0.188235            0.098039   
23           (ghee)         (sugar)            0.098039            0.188235   

     support  confidence      lift  leverage  conviction  
3   0.023529    0.666667  6.800000  0.020069    2.705882  
4   0.023529    0.315789  3.221053  0.016225    1.318250  
6   0.023529    0.375000  3.297414  0.016394    1.418039  
8   0.023529    0.315789  3.834586  0.017393    1.341176  
10  0.023529    0.600000  3.187500  0.016148    2.029412  
13  0.027451    0.777778  4.131944  0.020807    3.652941  
14  0.035294    0.360000  3.672000  0.025682    1.409314  
15  0.035294    0.360000  3.672000  0.025682    1.409314  
16  0.035294    0.333333  2.931034  0.023253    1.329412  
17  0.035294    0.310345  2.931034  0.023253    1.296471  
18  0.043137    0.687500  7.012500  0.036986    2.886275  
19  0.043137    0.440000  7.012500  0.036986    1.673669  
20  0.054902    0.666667  3.541667  0.039400    2.435294  
22  0.062745    0.333333  3.400000  0.044291    1.352941  
23  0.062745    0.640000  3.400000  0.044291    2.254902

a) Which are the three most common ingredients? (i.e. most common 1-itemsets)
- 0.188235	(sugar)
- 0.113725	(ginger)
- 0.105882	(garam masala)

In [163]:
a2_2  = freq_items_2[freq_items_2['length'] == 1].copy()
a2_2.nlargest(3,"support")

support        itemsets  length
45  0.188235         (sugar)       1
21  0.113725        (ginger)       1
17  0.105882  (garam masala)       1

b) What is the most common pair of ingredients (i.e. most common 2-itemsets)
- 0.062745	(sugar, ghee)

In [165]:
b2_2 = freq_items_2[freq_items_2['length'] == 2].copy()
b2_2.nlargest(1,'support')

support       itemsets  length
60  0.062745  (sugar, ghee)       2

c) What are your 3-5 most "interesting" rules?
With a minimum support value threshold of @% and a confidence threshold of 30%  The 5 most interesting rules we found
were:

- urad dal -> chana dal
- chana dal -> urad dal
- mustard seeds -> curry leaves
- caradamom -> sugar
- jaggery -> rice flour

These were the most interesting as they had the highest lift values

In [167]:
c2_1 = rules_2_min_conf_3.sort_values(["lift", "confidence"], ascending=False)
c2_1 = c2_1[c2_1['support'] >= .02]

c2_1.head(5)

antecedents     consequents  antecedent support  consequent support  \
18      (chana dal)      (urad dal)            0.062745            0.098039   
19       (urad dal)     (chana dal)            0.098039            0.062745   
3   (mustard seeds)  (curry leaves)            0.035294            0.098039   
13       (cardamom)         (sugar)            0.035294            0.188235   
8         (jaggery)    (rice flour)            0.074510            0.082353   

     support  confidence      lift  leverage  conviction  
18  0.043137    0.687500  7.012500  0.036986    2.886275  
19  0.043137    0.440000  7.012500  0.036986    1.673669  
3   0.023529    0.666667  6.800000  0.020069    2.705882  
13  0.027451    0.777778  4.131944  0.020807    3.652941  
8   0.023529    0.315789  3.834586  0.017393    1.341176